### PR TITLE
docs: profiling howto [DET-5209]

### PR DESCRIPTION
**Pull request by armandmcqueen**
## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: docs/how-to/profiling.txt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+-  Disk throughput (read)
+-  Disk throughput (write)
+-  Host memory sage
+-  Average CPU utilization across cores
+
+For distributed training, these metrics are collected for every agent.
+The data is broken down per agent. GPU metrics can be further broken
+down by GPU.
+
+Note: System Metrics record agent-level metrics, so when there are
```
Comment (stoksc) [inline docs/how-to/profiling.txt:42]: I'd use a
```
.. note::
```
block

File: docs/how-to/profiling.txt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+Aggregation
+
+*********
+ Timings
+*********
+
+The other class of profiling that Determined handles is Timings. This
+tracks how long training events take, such as retrieving data from the
+data loader, moving data host-to-device, running the forward/backwards
+pass, or executing callbacks.
```
Comment (stoksc) [inline docs/how-to/profiling.txt:61]:
```suggestion
The other type of profiling metric that Determined tracks are Timings. These
track how long training events take, such as retrieving data from the
data loader, moving data host-to-device, running the forward/backwards
pass, or executing callbacks.
```

File: docs/how-to/profiling.txt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+   takes to execute\*)
+-  ``step_lr_schedulers`` (amount time taken to update the LR schedules)
+-  ``from_device`` (amount of time transfering output from device to
+   host)
+-  ``reduce_metrics`` (amount of time taken to calculate global metrics
+   in distributed training)
+
+\* ``train_batch`` is typically the forward pass and the backwards pass.
+
+Timings are currently only supported in PyTorch.
```
Comment (stoksc) [inline docs/how-to/profiling.txt:81]: I may move this to the intro paragraph, or a note below it.

File: docs/how-to/profiling.txt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+tracks how long training events take, such as retrieving data from the
+data loader, moving data host-to-device, running the forward/backwards
+pass, or executing callbacks.
+
+This provides a fairly high-level picture of where you might want to
+spend your time further investigating the code.
+
+Specifically, Determined tracks:
+
+-  ``dataloader_next`` (amount of time to retrieve the next item from
```
Comment (stoksc) [inline docs/how-to/profiling.txt:68]:
```suggestion
-  ``dataloader_next`` (retrieving the next item from
```
and similar maybe?

File: docs/how-to/profiling.txt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+-  ``to_device`` (amount of time transfering input from host to device)
+-  ``train_batch`` (amount of time user-defined ``train_batch`` function
+   takes to execute\*)
+-  ``step_lr_schedulers`` (amount time taken to update the LR schedules)
+-  ``from_device`` (amount of time transfering output from device to
+   host)
+-  ``reduce_metrics`` (amount of time taken to calculate global metrics
+   in distributed training)
+
+\* ``train_batch`` is typically the forward pass and the backwards pass.
```
Comment (stoksc) [inline docs/how-to/profiling.txt:79]: typo?

File: docs/how-to/profiling.txt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+For distributed training, these metrics are collected for every agent.
+The data is broken down per agent. GPU metrics can be further broken
+down by GPU.
+
+Note: System Metrics record agent-level metrics, so when there are
+multiple experiments on the agent, it is difficult to analyze. We
+suggest that profiling is done when the experiment is the only thing
+running on the agent.
+
+*******************************
```
Comment (stoksc) [inline docs/how-to/profiling.txt:47]: I could take or leave this, or add it to a follow up if you wanted (seems it this could be a topic guide on profiling in determined and the other (this) could be a how to?).

File: docs/how-to/profiling.txt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ Example: Gradient Aggregation
+*******************************
+
+TODO: Walk through how to use System Metrics to look at Gradient
+Aggregation
+
+*********
+ Timings
+*********
```
Comment (stoksc) [inline docs/how-to/profiling.txt:56]: when @azhou-determined is done with the profiling tools work, maybe we could add a section here for "Native Tooling"

File: docs/how-to/profiling.txt [hunk 8]
```diff
@@ -0,0 +1,81 @@
+.. _howto-profiling:
+
+#################################
+ How To: Profile Your Experiment
+#################################
+
+When you want to improve the training speed of your model, the first
+thing you need to do is understand why training is slow. To do this, you
+need to profile your training code. There are many different layers you
+can look at, from raw throughput all the way to GPU kernals.
```
Comment (stoksc) [inline docs/how-to/profiling.txt:10]:
```suggestion
When you want to debug the training speed of a model, you need
tools to inspect and measure internal states and runtime statistics of
the model during execution. There are many different layers you
can look at, from raw throughput all the way to GPU kernels.
```
Idk this is a hard paragraph to write.. i want to take a course on technical writing. I feel like there is a better way to motivate the reader but i can't quite find the words

File: docs/how-to/profiling.txt [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+Determined provides two tools out-of-the-box for profiling your training
+- System Metrics and Timings.
```
[the remark this answers is not in the document]
Comment (armandmcqueen) [reply docs/how-to/profiling.txt:13]: Fixed

File: docs/how-to/profiling.txt (hunk 6 above)
Comment (armandmcqueen) [reply docs/how-to/profiling.txt:47]: I removed it. It can be an add on-later

File: docs/how-to/profiling.txt (hunk 2 above)
Comment (armandmcqueen) [reply docs/how-to/profiling.txt:61]: I improved this, although it's slightly different than this suggestion

File: docs/how-to/profiling.txt (hunk 4 above)
Comment (armandmcqueen) [reply docs/how-to/profiling.txt:68]: fixed

File: docs/how-to/profiling.txt (hunk 5 above)
Comment (armandmcqueen) [reply docs/how-to/profiling.txt:79]: How so?

File: docs/how-to/profiling.txt (hunk 3 above)
Comment (armandmcqueen) [reply docs/how-to/profiling.txt:81]: Fixed

File: docs/how-to/profiling.txt (hunk 8 above)
Comment (armandmcqueen) [reply docs/how-to/profiling.txt]: I rewrote it a bit. It could probably still use some improvements.

File: docs/how-to/profiling.txt [hunk 10]
```diff
@@ -0,0 +1,95 @@
+.. _howto-profiling:
+
+#################################
+ How To: Profile An Experiment
```
Comment (armandmcqueen) [inline docs/how-to/profiling.txt:4]: I think this page naming convention is useful, particularly for search, but it is inconsistent with the rest of the pages so maybe we should replace the title.

Comment (stoksc) [reply docs/how-to/profiling.txt:4]: you mean the prefix of "how to"?

Comment (armandmcqueen) [reply docs/how-to/profiling.txt:4]: Yeah

File: docs/how-to/profiling.txt [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+utilization and network throughput. These metrics are useful for seeing
+whether training is using the hardware effectively. When the System
+Metrics reported for an experiment are below what is expected from the
+hardware, that is a sign that the software may be able to be optimized
+to make better use of the hardware resources.
+
+Specifically, Determined tracks:
+
+-  GPU utilization
+-  GPU memory usage
```
Comment (stoksc) [inline docs/how-to/profiling.txt:47]: should we say free mem here and below? or like `GPU memory usage (as the amount of free GPU memory)`

**Review by stoksc — APPROVED** [review]
lgtm

File: docs/how-to/profiling.txt (hunk 11 above)
Comment (armandmcqueen) [reply docs/how-to/profiling.txt]: I updated it to be more accurate - "GPU free memory" and "Host available memory"
